### DISCUSSION TITLE
Transcription

### DISCUSSION THREAD

**Discussion by mattdrum** [chuot/rdio-scanner#409] [Ideas]
I am interested in the possibility of adding a speech to text feature in rdio-scanner for people who are hearing impaired. The two packages that I have played with are Whisper and Project Deepspeech. Both are open source project that run locally on your machine. An additional benefit of this is that audio files could then be searched based on keywords.

**Comment by BITBANSHEE-C137** [comment]
If you are savvy with LLM's, then whisper will work. not the fastest on micro hardware. Just add gas to make it go faster.

here are basic instruction on how to get started with Whisper, and a script to start up and watch your recordings. (modify to your own environment). 

NOTE: if Using RTL_AIRBAND as source, you will need to make a second output directory as rdio-scanner deletes the mp3 after ingestion into it's local db (default is sqlite)

Run the following commands in the Linux terminal to install Python, FFmpeg, and Whisper:

# Update package list
sudo apt update && sudo apt upgrade -y

# Install Python and dependencies
sudo apt install -y python3 python3-pip ffmpeg

# Install PyTorch for CPU (no GPU support needed)
pip3 install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cpu

# Install Whisper
pip3 install openai-whisper

# Verify installation
whisper --help
####################################################
#whisper_start.sh
#This version fixes decoding issues and ensures clean WAV conversion.
####################################################
#!/bin/bash

WATCH_DIR="/recordings/whisperin"
OUTPUT_DIR="/recordings/transcribed"
LOG_FILE="$OUTPUT_DIR/processed.log"
MODEL="small"  # Change model as needed

# Ensure the output directory exists
mkdir -p "$OUTPUT_DIR"

# Activate virtual environment
source ~/whisper_env/venv/bin/activate

# Create log file if it doesn't exist
touch "$LOG_FILE"

echo "Processing existing MP3 files in $WATCH_DIR..."
for FILE in "$WATCH_DIR"/*.mp3; do
    [ -e "$FILE" ] || continue  # Skip if no files exist

    # Check if file is already processed
    if grep -Fxq "$(basename "$FILE")" "$LOG_FILE"; then
        echo "Skipping already processed file: $FILE"
        continue
    fi

    echo "Processing: $FILE"

    # Convert MP3 to WAV format with error handling
    WAV_FILE="${FILE%.mp3}.wav"
    if ffmpeg -y -i "$FILE" -acodec pcm_s16le -ar 16000 -ac 1 -f wav "$WAV_FILE" 2>/dev/null; then
        echo "WAV conversion successful: $WAV_FILE"
    else
        echo "⚠️  Error converting MP3 to WAV: $FILE" >> "$OUTPUT_DIR/error.log"
        continue
    fi

    # Run Whisper transcription
    if whisper "$WAV_FILE" --model "$MODEL" --output_dir "$OUTPUT_DIR"; then
        echo "$(basename "$FILE")" >> "$LOG_FILE"
        echo "Transcription saved to $OUTPUT_DIR"
    else
        echo "⚠️  Error transcribing file: $FILE" >> "$OUTPUT_DIR/error.log"
    fi
done

echo "Watching $WATCH_DIR for new MP3 files..."

# Watch for new files
inotifywait -m -e close_write --format "%w%f" "$WATCH_DIR" | while read FILE
do
    if [[ "$FILE" == *.mp3 ]]; then

        # Check if file is already processed
        if grep -Fxq "$(basename "$FILE")" "$LOG_FILE"; then
            echo "Skipping already processed file: $FILE"
            continue
        fi

        echo "Processing: $FILE"

        # Convert MP3 to WAV format with error handling
        WAV_FILE="${FILE%.mp3}.wav"
        if ffmpeg -y -i "$FILE" -acodec pcm_s16le -ar 16000 -ac 1 -f wav "$WAV_FILE" 2>/dev/null; then
            echo "WAV conversion successful: $WAV_FILE"
        else
            echo "⚠️  Error converting MP3 to WAV: $FILE" >> "$OUTPUT_DIR/error.log"
            continue
        fi

        # Run Whisper transcription
        if whisper "$WAV_FILE" --model "$MODEL" --output_dir "$OUTPUT_DIR"; then
            echo "$(basename "$FILE")" >> "$LOG_FILE"
            echo "Transcription saved to $OUTPUT_DIR"
        else
            echo "⚠️  Error transcribing file: $FILE" >> "$OUTPUT_DIR/error.log"
        fi
    fi
done
